base_external_dbsource/tests/test_base_external_dbsource.py

```diff
@@ -0,0 +1,247 @@
+        with self.assertRaises(ConnectionSuccessError):
+            self.dbsource.connection_test()
+
+    def test_connection_fail(self):
+        """ It should raise for failed/invalid connection """
+        with mock.patch.object(self.dbsource, 'connection_open') as conn:
+            conn.side_effect = Exception
+            with self.assertRaises(ConnectionFailedError):
+                self.dbsource.connection_test()
+
+    def test_connection_open_calls_close(self):
+        """ It should close connection after context ends """
+        with mock.patch.object(
+            self.dbsource, 'connection_close',
+        ) as close:
+            with self.dbsource.connection_open():
+                pass
+            close.assert_called_once()
+
+    def test_connection_close(self):
+        """ It should call adapter's close method """
+        args = [mock.MagicMock()]
+        res, adapter = self._test_adapter_method(
+            'connection_close', args=args,
+        )
+        adapter.assert_called_once_with(args[0])
+
+    def test_execute_asserts_query_arg(self):
+        """ It should raise a TypeError if query and sqlquery not in args """
+        with self.assertRaises(TypeError):
+            self.dbsource.execute()
+
+    def test_execute_calls_adapter(self):
+        """ It should call the adapter methods with proper args """
+        expect = ('query', 'execute', 'metadata')
+        return_value = 'rows', 'cols'
+        res, adapter = self._test_adapter_method(
+            'execute', args=expect, return_value=return_value,
+        )
+        adapter.assert_called_once_with(*expect)
+
+    def test_execute_return(self):
+        """ It should return rows if not metadata """
+        expect = (True, True, False)
+        return_value = 'rows', 'cols'
+        res, adapter = self._test_adapter_method(
+            'execute', args=expect, return_value=return_value,
+        )
+        self.assertEqual(res, return_value[0])
+
+    def test_execute_return_metadata(self):
+        """ It should return rows and cols if metadata """
+        expect = (True, True, True)
+        return_value = 'rows', 'cols'
+        res, adapter = self._test_adapter_method(
+            'execute', args=expect, return_value=return_value,
+        )
+        self.assertEqual(
+            res,
+            {'rows': return_value[0],
+             'cols': return_value[1]},
+        )
+
+    def test_remote_browse(self):
+        """ It should call the adapter method with proper args """
+        args = [1], 'args'
+        kwargs = {'kwargs': True}
+        self.dbsource.current_table = 'table'
+        res, adapter = self._test_adapter_method(
+            'remote_browse', create=True, args=args, kwargs=kwargs,
+        )
+        adapter.assert_called_once_with(*args, **kwargs)
+        self.assertEqual(res, adapter())
+
+    def test_remote_browse_asserts_current_table(self):
+        """ It should raise AssertionError if a table not selected """
```

Comment:
(you can save yourself some effort and just tell me to apply X change to all instances)